### PR TITLE
ROX-12037: Support Quay Robot Accounts for registry integration

### PR DESCRIPTION
**Pull request by theencee**
## Description

When integrating Quay registry you can now optionally use robot account instead of just OAuth tokens. In fact this is Quay's recommended integration credentials. Unfortunately scanner still requires a token. 

Existing integrations should remain as-is.

## Checklist
- [ ] Investigated and inspected CI test results
- [x] Unit test and regression tests added
- [x] Evaluated and added CHANGELOG entry if required
~[ ] Determined and documented upgrade steps
- [ ] Documented user facing changes (create PR based on [openshift/openshift-docs](https://github.com/openshift/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))

If any of these don't apply, please comment below.

## Testing Performed

Auomated: See files
Manual:

Validated that on upgrade the existing integrations remain as is (with or without token).

Then just using the /test endpoint:
Registry:

1. No Auth - should pass OK

```
$ roxcurl v1/imageintegrations/test -X POST --data-raw '{"id":"","name":"quay test","categories":["REGISTRY"],"quay":{"endpoint":"quay.io","insecure":false},"autogenerated":false,"clusterId":"","clusters":[],"skipTestIntegration":false,"type":"quay"}' -sk | jq .
{}
```

2. Token Only - should pass OK

```
$ roxcurl v1/imageintegrations/test -X POST --data-raw '{"id":"","name":"quay test","categories":["REGISTRY"],"quay":{"endpoint":"quay.io","oauthToken":"REDACTED","insecure":false},"autogenerated":false,"clusterId":"","clusters":[],"skipTestIntegration":false,"type":"quay"}' -sk | jq .
{}
```

3. Robot Only - should pass OK

```
$ roxcurl v1/imageintegrations/test -X POST --data-raw '{"id":"","name":"quay test","categories":["REGISTRY"],"quay":{"endpoint":"quay.io","registryRobotCredentials":{"username":"REDACTED","password":"REDACTED"},"insecure":false},"autogenerated":false,"clusterId":"","clusters":[],"skipTestIntegration":false,"type":"quay"}' -sk | jq .
{}
```

4. Token + Robot - should fail OK

```
$ roxcurl v1/imageintegrations/test -X POST --data-raw '{"id":"","name":"quay test","categories":["REGISTRY"],"quay":{"endpoint":"quay.io","oauthToken":"REDACTED","registryRobotCredentials":{"username":"REDACTED","password":"REDACTED"},"insecure":false},"autogenerated":false,"clusterId":"","clusters":[],"skipTestIntegration":false,"type":"quay"}' -sk | jq .
{
  "error": "registry integration: Quay registry integration should use robot credentials or robot credentials but not both: invalid arguments: invalid arguments",
  "code": 3,
  "message": "registry integration: Quay registry integration should use robot credentials or robot credentials but not both: invalid arguments: invalid arguments",
  "details": []
}
```

5. Robot missing username - should fail OK

```
$ roxcurl v1/imageintegrations/test -X POST --data-raw '{"id":"","name":"quay test","categories":["REGISTRY"],"quay":{"endpoint":"quay.io","registryRobotCredentials":{"password":"REDACTED"},"insecure":false},"autogenerated":false,"clusterId":"","clusters":[],"skipTestIntegration":false,"type":"quay"}' -sk | jq .
{
  "error": "registry integration: If using Quay robot credentials, both username and password must be provided: invalid arguments: invalid arguments",
  "code": 3,
  "message": "registry integration: If using Quay robot credentials, both username and password must be provided: invalid arguments: invalid arguments",
  "details": []
}
```

6. Robot missing password - should fail OK

```
$ roxcurl v1/imageintegrations/test -X POST --data-raw '{"id":"","name":"quay test","categories":["REGISTRY"],"quay":{"endpoint":"quay.io","registryRobotCredentials":{"username":"REDACTED"},"insecure":false},"autogenerated":false,"clusterId":"","clusters":[],"skipTestIntegration":false,"type":"quay"}' -sk | jq .
{
  "error": "registry integration: If using Quay robot credentials, both username and password must be provided: invalid arguments: invalid arguments",
  "code": 3,
  "message": "registry integration: If using Quay robot credentials, both username and password must be provided: invalid arguments: invalid arguments",
  "details": []
}
```

Scanner:

1. No Auth - should pass OK

```
$ roxcurl v1/imageintegrations/test -X POST --data-raw '{"id":"","name":"quay test","categories":["SCANNER"],"quay":{"endpoint":"quay.io","insecure":false},"autogenerated":false,"clusterId":"","clusters":[],"skipTestIntegration":false,"type":"quay"}' -sk | jq .
{}
```

2. Token Only - should pass OK

```
$ roxcurl v1/imageintegrations/test -X POST --data-raw '{"id":"","name":"quay test","categories":["SCANNER"],"quay":{"endpoint":"quay.io","oauthToken":"REDACTED","insecure":false},"autogenerated":false,"clusterId":"","clusters":[],"skipTestIntegration":false,"type":"quay"}' -sk | jq .
{}
```

3. Robot Only - should fail OK

```
$ roxcurl v1/imageintegrations/test -X POST --data-raw '{"id":"","name":"quay test","categories":["SCANNER"],"quay":{"endpoint":"quay.io","registryRobotCredentials":{"username":"REDACTED","password":"REDACTED"},"insecure":false},"autogenerated":false,"clusterId":"","clusters":[],"skipTestIntegration":false,"type":"quay"}' -sk | jq .
{
  "error": "image scanner integration: Quay scanner integration cannot use robot credentials: invalid arguments: invalid arguments",
  "code": 3,
  "message": "image scanner integration: Quay scanner integration cannot use robot credentials: invalid arguments: invalid arguments",
  "details": []
}
```

4. Token + Robot - should fail 

```
$ roxcurl v1/imageintegrations/test -X POST --data-raw '{"id":"","name":"quay test","categories":["SCANNER"],"quay":{"endpoint":"quay.io","oauthToken":"REDACTED","registryRobotCredentials":{"username":"REDACTED","password":"REDACTED"},"insecure":false},"autogenerated":false,"clusterId":"","clusters":[],"skipTestIntegration":false,"type":"quay"}' -sk | jq .
{
  "error": "image scanner integration: Quay scanner integration cannot use robot credentials: invalid arguments: invalid arguments",
  "code": 3,
  "message": "image scanner integration: Quay scanner integration cannot use robot credentials: invalid arguments: invalid arguments",
  "details": []
}

```

5. Robot missing username - should fail 

```
$ roxcurl v1/imageintegrations/test -X POST --data-raw '{"id":"","name":"quay test","categories":["SCANNER"],"quay":{"endpoint":"quay.io","registryRobotCredentials":{"password":"REDACTED"},"insecure":false},"autogenerated":false,"clusterId":"","clusters":[],"skipTestIntegration":false,"type":"quay"}' -sk | jq .
{
  "error": "image scanner integration: Quay scanner integration cannot use robot credentials: invalid arguments: invalid arguments",
  "code": 3,
  "message": "image scanner integration: Quay scanner integration cannot use robot credentials: invalid arguments: invalid arguments",
  "details": []
}
```

6. Robot missing password - should fail OK

```
$ roxcurl v1/imageintegrations/test -X POST --data-raw '{"id":"","name":"quay test","categories":["SCANNER"],"quay":{"endpoint":"quay.io","registryRobotCredentials":{"username":"REDACTED"},"insecure":false},"autogenerated":false,"clusterId":"","clusters":[],"skipTestIntegration":false,"type":"quay"}' -sk | jq .
{
  "error": "image scanner integration: Quay scanner integration cannot use robot credentials: invalid arguments: invalid arguments",
  "code": 3,
  "message": "image scanner integration: Quay scanner integration cannot use robot credentials: invalid arguments: invalid arguments",
  "details": []
}

```

Registry + Scanner:

1. No Auth - should pass 

```
$ roxcurl v1/imageintegrations/test -X POST --data-raw '{"id":"","name":"quay test","categories":["REGISTRY","SCANNER"],"quay":{"endpoint":"quay.io","insecure":false},"autogenerated":false,"clusterId":"","clusters":[],"skipTestIntegration":false,"type":"quay"}' -sk | jq .
{}
```

2. Token Only - should pass OK

```
$ roxcurl v1/imageintegrations/test -X POST --data-raw '{"id":"","name":"quay test","categories":["REGISTRY","SCANNER"],"quay":{"endpoint":"quay.io","oauthToken":"REDACTED","insecure":false},"autogenerated":false,"clusterId":"","clusters":[],"skipTestIntegration":false,"type":"quay"}' -sk | jq .
{}
```

3. Robot Only - should fail OK

```
$ roxcurl v1/imageintegrations/test -X POST --data-raw '{"id":"","name":"quay test","categories":["REGISTRY","SCANNER"],"quay":{"endpoint":"quay.io","registryRobotCredentials":{"username":"REDACTED","password":"REDACTED"},"insecure":false},"autogenerated":false,"clusterId":"","clusters":[],"skipTestIntegration":false,"type":"quay"}' -sk | jq .
{
  "error": "registry integration: Quay scanner integration cannot use robot credentials: invalid arguments: invalid arguments",
  "code": 3,
  "message": "registry integration: Quay scanner integration cannot use robot credentials: invalid arguments: invalid arguments",
  "details": []
}
```

4. Token + Robot - should pass OK

```
$ roxcurl v1/imageintegrations/test -X POST --data-raw '{"id":"","name":"quay test","categories":["REGISTRY","SCANNER"],"quay":{"endpoint":"quay.io","oauthToken":"REDACTED","registryRobotCredentials":{"username":"REDACTED","password":"REDACTED"},"insecure":false},"autogenerated":false,"clusterId":"","clusters":[],"skipTestIntegration":false,"type":"quay"}' -sk | jq .
{}
```

5. Robot missing username - should fail OK

```
$ roxcurl v1/imageintegrations/test -X POST --data-raw '{"id":"","name":"quay test","categories":["REGISTRY","SCANNER"],"quay":{"endpoint":"quay.io","registryRobotCredentials":{"password":"REDACTED"},"insecure":false},"autogenerated":false,"clusterId":"","clusters":[],"skipTestIntegration":false,"type":"quay"}' -sk | jq .
{
  "error": "registry integration: Quay scanner integration cannot use robot credentials: invalid arguments: invalid arguments",
  "code": 3,
  "message": "registry integration: Quay scanner integration cannot use robot credentials: invalid arguments: invalid arguments",
  "details": []
}
```

6. Robot missing password - should fail OK

```
$ roxcurl v1/imageintegrations/test -X POST --data-raw '{"id":"","name":"quay test","categories":["REGISTRY","SCANNER"],"quay":{"endpoint":"quay.io","registryRobotCredentials":{"username":"REDACTED"},"insecure":false},"autogenerated":false,"clusterId":"","clusters":[],"skipTestIntegration":false,"type":"quay"}' -sk | jq .
{
  "error": "registry integration: Quay scanner integration cannot use robot credentials: invalid arguments: invalid arguments",
  "code": 3,
  "message": "registry integration: Quay scanner integration cannot use robot credentials: invalid arguments: invalid arguments",
  "details": []
}
```
And then by actually creating some integrations and using it:

1. Add registry with token only

```
$ roxcurl v1/imageintegrations -X POST --data-raw '{"id":"","name":"quay test","categories":["REGISTRY"],"quay":{"endpoint":"quay.io","oauthToken":"REDACTED","insecure":false},"autogenerated":false,"clusterId":"","clusters":[],"skipTestIntegration":false,"type":"quay"}' -sk | jq .
{
  "id": "94dca201-6f91-4396-bca5-f48be015e27d",
  "name": "quay test",
  "type": "quay",
  "categories": [
    "REGISTRY"
  ],
  "quay": {
    "endpoint": "quay.io",
    "oauthToken": "REDACTED",
    "insecure": false,
    "registryRobotCredentials": null
  },
  [..snip..]
}

$ roxctl image scan -i quay.io/rhacs-eng/qa:struts-app -f | jq '.metadata.dataSource, .metadata.v1.layers[0], .scan.components[0].name, .scan.dataSource'
{
  "id": "94dca201-6f91-4396-bca5-f48be015e27d",
  "name": "quay test"
}
{
  "instruction": "ADD",
  "value": "file:9c48682ff75c756544d4491472081a078edf5dd0bb5038d1cb850a1f9c480e3e in /",
  "created": "2017-06-20T20:13:32.945602158Z"
}
"readline6"
{
  "id": "169b0d3f-8277-4900-bbce-1127077defae",
  "name": "Stackrox Scanner"
}
```

2. Add registry with robot only

```
$ roxcurl v1/imageintegrations -X POST --data-raw '{"id":"","name":"quay test","categories":["REGISTRY"],"quay":{"endpoint":"quay.io","registryRobotCredentials":{"username":"REDACTED","password":"REDACTED"},"insecure":false},"autogenerated":false,"clusterId":"","clusters":[],"skipTestIntegration":false,"type":"quay"}' -sk | jq .
{
  "id": "69b8a1d5-289b-428b-b6ec-3d11d7a7247c",
  "name": "quay test",
  "type": "quay",
  "categories": [
    "REGISTRY"
  ],
  "quay": {
    "endpoint": "quay.io",
    "oauthToken": "",
    "insecure": false,
    "registryRobotCredentials": {
      "username": "REDACTED",
      "password": "REDACTED"
    }
  },
  [..snip..]
}

$ roxctl image scan -i quay.io/rhacs-eng/qa:struts-app -f | jq '.metadata.dataSource, .metadata.v1.layers[0], .scan.components[0].name, .scan.dataSource'
{
  "id": "69b8a1d5-289b-428b-b6ec-3d11d7a7247c",
  "name": "quay test"
}
{
  "instruction": "ADD",
  "value": "file:9c48682ff75c756544d4491472081a078edf5dd0bb5038d1cb850a1f9c480e3e in /",
  "created": "2017-06-20T20:13:32.945602158Z"
}
"libvorbis"
{
  "id": "169b0d3f-8277-4900-bbce-1127077defae",
  "name": "Stackrox Scanner"
}
```

3. Add registry + scanner with token only

```
$ roxcurl v1/imageintegrations -X POST --data-raw '{"id":"","name":"quay test","categories":["REGISTRY","SCANNER"],"quay":{"endpoint":"quay.io","oauthToken":"REDACTED","insecure":false},"autogenerated":false,"clusterId":"","clusters":[],"skipTestIntegration":false,"type":"quay"}' -sk | jq .
{
  "id": "226ed458-45db-4c7b-afe0-be52f0ffefef",
  "name": "quay test",
  "type": "quay",
  "categories": [
    "REGISTRY",
    "SCANNER"
  ],
  "quay": {
    "endpoint": "quay.io",
    "oauthToken": "REDACTED",
    "insecure": false,
    "registryRobotCredentials": null
  },
  [..snip..]
}

$ roxctl image scan -i quay.io/rhacs-eng/qa:struts-app -f | jq '.metadata.dataSource, .metadata.v1.layers[0], .scan.components[0].name, .scan.dataSource'
{
  "id": "226ed458-45db-4c7b-afe0-be52f0ffefef",
  "name": "quay test"
}
{
  "instruction": "ADD",
  "value": "file:9c48682ff75c756544d4491472081a078edf5dd0bb5038d1cb850a1f9c480e3e in /",
  "created": "2017-06-20T20:13:32.945602158Z"
}
"libogg0"
{
  "id": "226ed458-45db-4c7b-afe0-be52f0ffefef",
  "name": "quay test"
}
```

4. Add registry + scanner with token & robot

```
$ roxcurl v1/imageintegrations -X POST --data-raw '{"id":"","name":"quay test","categories":["REGISTRY","SCANNER"],"quay":{"endpoint":"quay.io","oauthToken":"REDACTED","registryRobotCredentials":{"username":"REDACTED","password":"REDACTED"},"insecure":false},"autogenerated":false,"clusterId":"","clusters":[],"skipTestIntegration":false,"type":"quay"}' -sk | jq .
{
  "id": "5e750920-a650-4fe6-9973-e5ca765f3984",
  "name": "quay test",
  "type": "quay",
  "categories": [
    "REGISTRY",
    "SCANNER"
  ],
  "quay": {
    "endpoint": "quay.io",
    "oauthToken": "REDACTED",
    "insecure": false,
    "registryRobotCredentials": {
      "username": "REDACTED",
      "password": "REDACTED"
    }
  },
  [..snip..]
}

$ roxctl image scan -i quay.io/rhacs-eng/qa:struts-app -f | jq '.metadata.dataSource, .metadata.v1.layers[0], .scan.components[0].name, .scan.dataSource'
{
  "id": "5e750920-a650-4fe6-9973-e5ca765f3984",
  "name": "quay test"
}
{
  "instruction": "ADD",
  "value": "file:9c48682ff75c756544d4491472081a078edf5dd0bb5038d1cb850a1f9c480e3e in /",
  "created": "2017-06-20T20:13:32.945602158Z"
}
"org.apache.myfaces.core:myfaces-impl"
{
  "id": "5e750920-a650-4fe6-9973-e5ca765f3984",
  "name": "quay test"
}
```

### CODE REVIEW

File: pkg/registries/quay/quay.go [hunk 1]
```diff
@@ -35,28 +36,70 @@ type Quay struct {
...
+			// If scanner only, robot credentials doesn't work. So expect either OAuth token or nothing (public registry)
+			if quay.GetRegistryRobotCredentials() != nil {
+				return errors.New("Quay scanner integration cannot use robot credentials")
+			}
+		} else if len(categories) == 1 && categories[0] == storage.ImageIntegrationCategory_REGISTRY {
+			// If registry only, only one of OAuth token, robot credentials or neither is allowed. Error if both are provided
+			if quay.GetRegistryRobotCredentials() != nil && quay.GetOauthToken() != "" {
+				return errors.New("Quay registry integration should use robot credentials or robot credentials but not both")
+			}
+		} else {
```
Comment (RTann) [inline pkg/registries/quay/quay.go:55]: is it possible for `len(categories) == 0` here? I think being inside this `else` assumes `len(categories) == 2`, right?

File: pkg/registries/quay/quay.go [hunk 2]
```diff
@@ -35,28 +36,70 @@ type Quay struct {
...
+	if features.QuayRobotAccounts.Enabled() {
+		if len(categories) == 1 && categories[0] == storage.ImageIntegrationCategory_SCANNER {
+			// If scanner only, robot credentials doesn't work. So expect either OAuth token or nothing (public registry)
+			if quay.GetRegistryRobotCredentials() != nil {
+				return errors.New("Quay scanner integration cannot use robot credentials")
+			}
+		} else if len(categories) == 1 && categories[0] == storage.ImageIntegrationCategory_REGISTRY {
+			// If registry only, only one of OAuth token, robot credentials or neither is allowed. Error if both are provided
+			if quay.GetRegistryRobotCredentials() != nil && quay.GetOauthToken() != "" {
+				return errors.New("Quay registry integration should use robot credentials or robot credentials but not both")
```
Comment (RTann) [inline pkg/registries/quay/quay.go:53]: nit: should use either OAuth or robot credentials but not both (robot is listed twice)

File: pkg/registries/quay/quay.go [hunk 3]
```diff
@@ -35,28 +36,70 @@ type Quay struct {
...
+			// Both being empty is ok as that's a public registry.
+			if quay.GetOauthToken() == "" && quay.GetRegistryRobotCredentials() != nil {
+				return errors.New("Quay scanner integration cannot use robot credentials")
+			}
+		}
+
+		// If using robot creds, check that both username and password is provided.
+		if quay.GetRegistryRobotCredentials() != nil {
+			if quay.GetRegistryRobotCredentials().GetUsername() == "" || quay.GetRegistryRobotCredentials().GetPassword() == "" {
+				return errors.New("If using Quay robot credentials, both username and password must be provided")
```
Comment (RTann) [inline pkg/registries/quay/quay.go:67]: nit: maybe phrase this like "Both username and password must be provided when using Quay robot credentials"

File: pkg/registries/quay/quay.go [hunk 4]
```diff
@@ -35,28 +36,70 @@ type Quay struct {
...
+			}
+		}
+	}
+
 	// Note that the oauth token could be empty because there are public images
```
Comment (RTann) [inline pkg/registries/quay/quay.go:72]: should we include robot in this comment, too?

File: pkg/registries/quay/quay.go [hunk 5]
```diff
@@ -35,28 +36,70 @@ type Quay struct {
...
 
-	var username string
-	if config.GetOauthToken() != "" {
-		username = oauthTokenString
+	var username, password string
+	password = config.GetOauthToken()
+
+	if features.QuayRobotAccounts.Enabled() {
+		if config.GetRegistryRobotCredentials() != nil {
+			// If robot credentials are provided use it for registry, regardless of if ImageIntegration is also for scanner.
```
Comment (RTann) [inline pkg/registries/quay/quay.go:87]: nit: regardless of whether ImageIntegration is also for scanner

File: pkg/registries/quay/quay.go [hunk 6]
```diff
@@ -35,28 +36,70 @@ type Quay struct {
...
+			// The scanner portion of it can use OAuth token, but the registry object should use proper robot creds.
+			username = config.GetRegistryRobotCredentials().GetUsername()
+			password = config.GetRegistryRobotCredentials().GetPassword()
+		} else if config.GetOauthToken() != "" {
+			username = oauthTokenString
```
Comment (RTann) [inline pkg/registries/quay/quay.go:92]: nit: maybe just always do this check like before and just overwrite it with robot creds if provided? I think that will remove the need for these `else`s

File: pkg/registries/quay/quay_test.go [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+	"github.com/stackrox/rox/pkg/features"
+	"github.com/stackrox/rox/pkg/testutils/envisolator"
+	"github.com/stretchr/testify/assert"
+)
+
+func TestValidate(t *testing.T) {
+	envIsolator := envisolator.NewEnvIsolator(t)
+	envIsolator.Setenv(features.QuayRobotAccounts.EnvVar(), "true")
+
+	if !features.QuayRobotAccounts.Enabled() {
```
Comment (RTann) [inline pkg/registries/quay/quay_test.go:16]: You can use https://github.com/stackrox/stackrox/blob/master/pkg/testutils/feature_flag.go#L24 if you want

**Review by RTann — APPROVED** [review]
mostly nits otherwise LGTM

File: pkg/registries/quay/quay.go (hunk 1 above)
Comment (theencee) [reply pkg/registries/quay/quay.go:55]: It gets validated elsewhere

**Pull request merged**